journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+package com.evolutiongaming.kafka.journal.util
+
+import cats.instances.all._
+import cats.effect.implicits._
+import cats.effect.kernel.{Async, Deferred, Sync}
+import cats.effect.std.Mutex
+import cats.effect.{Ref, Resource}
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.Log
+
+import java.util.UUID
+import java.util.concurrent.TimeoutException
+import scala.concurrent.duration.FiniteDuration
+
+private[journal] trait ResourcePool[F[_], O] {
+
+  /**
+   * `acquire` function of the Resource semantically blocks if there are no objects available.
+   *
+   * `release` function of the Resource returns the object to the pool.
+   * Note that it will NOT deallocate the object.
+   */
+  def borrow: Resource[F, O]
+}
+
+private[journal] object ResourcePool {
+
+  type ObjectId = UUID
+  type WaiterId = UUID
+  type Waiter[F[_], O] = Deferred[F, Either[Throwable, (ObjectId, O)]]
+
+  case class AllocatedObject[F[_], O](value: O, deallocate: F[Unit], state: AllocatedObject.State)
+
+  object AllocatedObject {
+    sealed abstract class State(val isFree: Boolean)
+
+    object State {
+      case object Used extends State(false)
+      case class Free(releasedAt: FiniteDuration) extends State(true)
+    }
+  }
+
+  case class State[F[_], O](
+    objects: Map[ObjectId, AllocatedObject[F, O]],
+    waiters: Map[WaiterId, Waiter[F, O]],
+  ) {
+
+    def findFreeObject: Option[(ObjectId, AllocatedObject[F, O])] =
+      objects.collectFirst {
+        case (id, obj) if obj.state.isFree => (id, obj)
+      }
+
+    def markObjectAsUsed(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Used)),
+        waiters
+      )
+
+    def markObjectAsFree(id: ObjectId, obj: AllocatedObject[F, O], releasedAt: FiniteDuration): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Free(releasedAt))),
+        waiters
+      )
+
+    def addObject(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj),
+        waiters
+      )
+
+    def removeObject(id: ObjectId): State[F, O] =
+      State(
+        objects - id,
+        waiters
+      )
+
+    def addWaiter(id: WaiterId, waiter: Waiter[F, O]): State[F, O] =
+      State(
+        objects,
+        waiters + (id -> waiter),
+      )
+
+    def removeWaiter(id: WaiterId): State[F, O] =
+      State(
+        objects,
+        waiters - id
+      )
+  }
+
+  def fixedSize[F[_]: Async, O](
```

Comment:
not sure you can call it a `fixedSize` now

Reply:
I think this is still fine 😄 
![Screenshot 2023-09-04 at 11 24 47](https://github.com/evolution-gaming/kafka-journal/assets/5853654/576d8370-2f3d-466b-a67a-6606ded1f51c)